tools/layoutflash/src/main.rs

```diff
@@ -1,46 +1,53 @@
 use clap::Parser;
-use device_tree::area::{read_areas, Area};
 use std::io::{self, Seek, SeekFrom, Write};
 use std::process::exit;
 use std::{
     env, fs,
     path::{Path, PathBuf},
 };
 
-fn read_fixed_fdt(path: &Path) -> io::Result<Vec<Area>> {
-    let data = match fs::read(path) {
-        Err(e) => {
-            return Err(io::Error::new(
-                e.kind(),
-                format!("{}{}", "Could not open: ", path.display()),
-            ))
-        }
-        Ok(data) => data,
-    };
-    let driver = SliceReader::new(data.as_slice());
-
-    // no error returned from helper
-    let areas = read_areas(&driver).unwrap();
-
-    Ok(areas.to_vec())
+// NOTE: we don't use u32. At the rate that SPI flash is expanding, we're going to see
+// 5B addressing soon I bet. The size limitation should be a function of the destination,
+// not this program. This problem should just stupidly arrange things.
+struct Area {
+    pub name: String,
+    pub offset: Option<usize>,
+    pub size: usize,
+    pub file: Option<String>,
 }
 
-// This method assumes that areas are sorted by offset.
-fn layout_flash(path: &Path, areas: &mut [Area]) -> io::Result<()> {
-    areas.sort_unstable_by_key(|a| a.offset);
+// In earlier versions of this function, we assumed all Areas had a non-zero
+// offset. There was a sort step to sort by offset as a first step.
+// Requiring users to compute all the offsets, and adjust them every time
+// something changed size, was incredibly inconvenient.
+// Now, we allow a zero offset, but it turns out that sorting is the wrong
+// thing to do: users tend to specify the areas in the dts in the order
+// they need to be in ROM; further, the rule of "put this area here,
+// and everything after it, after it" will not work if we sort: all
+// offset 0 items will be placed first!
+// So, new rules:
+// This method assumes that the areas are specified in order for a reason.
+// In most cases, the offset is 0, meaning "use the previous area end+1
+// as our offset". In rare cases, an offset is specified; for that, ensure
+// there is no overlap with previous Areas. All Areas after this Area will
+// be placed after it. That way, only a limited number of offsets need
+// to be specified, possibly even 0, and the order in the ROM image will be the order
+// specified in the DTS.
+fn layout_flash(path: &Path, areas: &mut Vec<Area>) -> io::Result<()> {
```

Comment:
``
```suggestion
fn layout_flash(path: &Path, areas: &mut [Area]) -> io::Result<()> {
```

minor style, since no `Vec` specific functionality is used

Reply:
This will make the software more Rusty, I'd suggest we apply this commit